sui_programmability/examples/nft_standard/sources/safe.move

```diff
@@ -0,0 +1,108 @@
+        borrowed: VecSet<ID>,
+        /// Valid version for TransferCap's
+        transfer_cap_version: u64,
+        /// Valid version for BorrowCap's
+        borrow_cap_version: u64,
+        /// If true, anyone can place NFT's into this safe.
+        /// If false, deposits require the OwnerCap
+        accept_deposits: bool,
+    }
+
+    /// A unique capability held by the owner of a particular `Safe`.
+    /// The holder can issue and revoke `TransferCap`'s and `BorrowCap`'s.
+    /// Can be used an arbitrary number of times
+    struct OwnerCap has key, store {
+        id: UID,
+        /// The ID of the safe that this capability grants permissions to
+        safe_id: ID,
+        /// Version of this cap.
+        version: u64,
+    }
+
+    /// Gives the holder permission to transfer the nft with id `nft_id` out of
+    /// the safe with id `safe_id`. Can only be used once.
+    struct TransferCap has key, store {
+        id: UID,
+        /// The ID of the safe that this capability grants permissions to
+        safe_id: ID,
+        /// The ID of the NFT that this capability can transfer
+        nft_id: ID,
+        version: u64
+    }
+
+    /// Gives the holder permission to borrow the nft with id `nft_id` out of
+    /// the safe with id `safe_id`. Can be used an arbitrary number of times.
+    struct BorrowCap has key, store {
+        id: UID,
+        /// The ID of the safe that this capability grants permissions to
+        safe_id: ID,
+        /// The ID of the NFT that this capability can transfer
+        nft_id: ID,
+        version: u64
+    }
+
+    /// "Hot potato" wrapping the borrowed NFT. Must be returned to `safe_id`
+    /// before the end of the current transaction
+    struct Borrowed<T> {
+        nft: T,
+        /// The safe that this NFT came from
+        safe_id: ID,
+        /// If true, only an immutable reference to `nft` can be granted
+        /// Always false if the NFT is currently listed
+        is_mutable: bool,
+    }
+
+    /// Create and share a fresh Safe that can hold T's.
+    /// Return an `OwnerCap` for the Safe
+    public fun create<T>(ctx: &mut TxContext): OwnerCap {
+        abort(0)
+    }
+
+    /// Produce a `TransferCap` for the NFT with `id` in `safe`.
+    /// This `TransferCap` can be (e.g.) used to list the NFT on a marketplace.
+    public fun sell_nft<T>(owner_cap: &OwnerCap, id: ID, safe: &mut T): TransferCap {
+        abort(0)
+    }
+
+    /// Consume `cap`, remove the NFT with `id` from `safe`, and return it to the caller.
+    /// Requiring `royalty` ensures that the caller has paid the required royalty for this collection
+    /// before completing  the purchase.
+    /// This invalidates all other `TransferCap`'s by increasing safe.transfer_cap_version
+    public fun buy_nft<T>(cap: TransferCap, royalty: RoyaltyReceipt<T>, id: ID, safe: &mut Safe<T>): T {
```

Comment:
4. The `RoyaltyReceipt` type only works if the TSCs integrate with the `Safe` type.

4a. Correctness. If the `Safe` type is not integrated in the method which performs the trade, it's up to the client (who has to batch txs) to prevent a race condition where two users could obtain `TransferCap` at the same time. Unless the `safe::buy_nft` is called atomically with the trade (and fails tx if the NFT was already sold), then both users paid protocol fees but only one claimed NFT. We're not convinced this subtle problem should be left on the clients. Therefore, TSCs should integrate with `Safe`.

4b. Enforcement. By taking enforcement out of trading contracts and having it done via `Safe`, honest trading contracts are limited to always work with `Safe` because otherwise any client could just sell their NFT as an object.

4c. By working with `Safe`, clients must know _up front_ which exact `Safe` instance to provide into the entry method. That is not a problem for slowly trading TSCs. Clients can retry. However, frequent trades may render the OB become unusable.
  The client has to fetch the OB state to know what's the lowest ask, because that determines what `Safe` to include in a tx to create a new bid.
  The client then has to send the tx. If lowest ask changed, the tx fails. The client has to retry.
  The client is interested in any collection's NFT, yet it observes failures due to abstraction leak.

4d. The OB cannot give some sort of receipt and let the user claim the NFT later via its own endpoint, because of the race condition mentioned above.

Reply:
To elaborate on 4c: to lessen the frequency of collisions the client could provide an array of `Safe` objects, e.g. for the cheapest N NFTs. However, if someone happens to offer their NFT for the lowest price after the client observes the state of the OB, their tx would still fail.

Reply:
>  Correctness. If the Safe type is not integrated in the method which performs the trade, it's up to the client (who has to batch txs) to prevent a race condition where two users could obtain TransferCap at the same time. Unless the safe::buy_nft is called atomically with the trade (and fails tx if the NFT was already sold), then both users paid protocol fees but only one claimed NFT. We're not convinced this subtle problem should be left on the clients. Therefore, TSCs should integrate with Safe.

Unfortunately, I think this problem is fundamental if we want to support (a) listing on multiple marketplaces simultaneously and/or (b) revocation of `TransferCap`'s (and I think both of these are important--e.g., ERC721 allows this).

Note that the "loser" of the race will only lose the gas fees for the aborted tx, *not* the cost of the NFT + commission + royalty--what will happen is that `Safe::buy` will abort with an invalid capability error (and thus no payments or other effects happen). This is unfortunate, but this same sort of thing happens on Eth, and (as mentioned above) seems like a fundamental consequence of sharing the permission to transfer an NFT.

Reply:
That's my understanding too: contracts have to integrate with the `Safe` to make the "loss" tolerable, ie. only the gas fee. `Safe::buy` must be always called in the same tx as where the trade happens.

Therefore when a trading contract exposes its `public entry fun` for buying an NFT, it must accept a `Safe` object. This leads to the issues mentioned in 4c.

A compromise we're offering here is to enable `TransferCap` to be optionally exclusive. Therefore OB or other contracts where exclusivity matters can use the pattern of giving a user some sort of receipt redeemable for the NFT _later_ (e.g. when the corresponding `Safe` ID is known.) Yet other trading contracts such as auctions where one always knows which `Safe` to include could benefit from the multi-listing feature.

Reply:
Yep, I think the "UniqueTransferCap" (which should also be non-revocable) idea makes sense. This is definitely a useful abstraction.